lib/console/hardware_identifiers/hardware_identifier.ex

```diff
@@ -0,0 +1,42 @@
+defmodule Console.HardwareIdentifiers.HardwareIdentifier do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query, warn: false
+
+  alias Console.Repo
+  alias Console.Gateways.Gateway
+  alias Console.HardwareIdentifiers.HardwareIdentifier
+
+  @primary_key {:id, :binary_id, autogenerate: true}
+  @foreign_key_type :binary_id
+  schema "hardware_identifiers" do
+    field :token, :binary
+
+    has_one :gateway, Gateway, on_delete: :delete_all
+
+    timestamps()
+  end
+
+  def changeset(hardware_identifier) do
+    token = :crypto.strong_rand_bytes(4)
+
+    result = hardware_identifier
+    |> change(token: token)
+    |> validate_required([:token])
+    |> unique_constraint(:token)
+  end
+
+  def get_associated_resource(token, resource) do
+    table =
+      case resource do
+        :gateway -> Gateway
+      end
+
+    query = from h in HardwareIdentifier,
+      left_join: t in ^table, on: [hardware_identifier_id: h.id],
```

Comment:
if you passed in a straight ecto schema like `Gateway` for resource, rather than `:gateway`, could you use that directly in place of `^table`? (to avoid the case)

Reply:
niceeeeee

Reply:
also the statement reads a little backwards. you might want to start with `query = from r in resource` and then join to `HardwareIdentifier`. then you don't have to swap out the `select` at the end. usually you'd start with the thing you want to end up with